Slice voxel kernel api changes

sampath1117 opened the pull request:
- Modified the slice kernel to accept anchor and shape params
- Added support to fill the padded region with fill value when padding is enabled 
- Added test cases for slice in audio and voxel test suites
- Modify the non silent region kernel output types to make it compatible with latest slice

r-abishek commented:
@sampath1117 Conflicts
Please merge master into your branch

r-abishek commented on src/modules/hip/kernel/slice.hpp:
```diff
@@ -32,79 +87,487 @@ __global__ void slice_ncdhw_hip_tensor(T *srcPtr,
-        int globalThreads_x = (dstGenericDescPtr->strides[3] + 7) >> 3; // W - width (x direction) - vectorized for 8 element loads/stores per channel
-        int globalThreads_y = dstGenericDescPtr->dims[3];               // H - height (y direction)
-        int globalThreads_z = dstGenericDescPtr->dims[2];               // D - depth (z direction)
+        // NHW
```
NHW? Pls remove, or add equivalent comments in other if branches too

r-abishek commented on src/modules/hip/kernel/slice.hpp:
```diff
@@ -32,79 +87,487 @@ __global__ void slice_ncdhw_hip_tensor(T *srcPtr,
+}
+
+template <typename T>
+RppStatus hip_exec_fill_value_tensor(T *dstPtr,
```
Suggest to always place all dependencies above the function that calls those deps. Since this is a dependency for hip_exec_slice_tensor(), place it right above it.

sampath1117 replied:
Done
Again, it is somehow not showing here

r-abishek commented on src/modules/hip/kernel/slice.hpp:
```diff
@@ -32,79 +87,487 @@ __global__ void slice_ncdhw_hip_tensor(T *srcPtr,
+{
+    if(numDims == 4)
+    {
+        Rpp32u paramDims = numDims;
```
Same comment, pls remove extra variable

r-abishek commented on src/modules/hip/kernel/slice.hpp:
```diff
@@ -32,79 +87,487 @@ __global__ void slice_ncdhw_hip_tensor(T *srcPtr,
+            }
         }
     }
+    hipStreamSynchronize(handle.GetStream());
```
Pls move the stream sync outside. To the line after the call to hip_exec_fill_value_tensor()

r-abishek commented on src/modules/hip/kernel/slice.hpp:
```diff
@@ -2,24 +2,79 @@
+
+    uint dstIdx = (id_z * dstStridesCDH.y) + (id_y * dstStridesCDH.z) + id_x;
+    d_float8 val_f8;
+    val_f8.f4[0] = (float4)(*fillValue);
```
Instead of checking if fillValue is needed, and running a separate kernel like this, have we evaluated how the performance is on simply doing a hipMemset for the whole buffer?

sampath1117 replied:
hipMemset, memset only works well when the value we want to memset is 0
For other values, it is recommended. Below is one online reference i found for the same
https://stackoverflow.com/questions/51053705/initialize-float-type-array-using-memset

r-abishek commented on src/modules/cpu/kernel/slice.hpp:
```diff
@@ -26,184 +26,210 @@ SOFTWARE.
-    else if(srcGenericDescPtr->layout==RpptLayout::NDHWC)
-        roiDefault = {0, 0, 0, (Rpp32s)srcGenericDescPtr->dims[3], (Rpp32s)srcGenericDescPtr->dims[2], (Rpp32s)srcGenericDescPtr->dims[1]};
     Rpp32u numThreads = handle.GetNumThreads();
+    Rpp32u numDims = srcGenericDescPtr->numDims - 1;
```
Same here - add a comment as to why this is done

r-abishek commented on src/modules/cpu/kernel/slice.hpp:
```diff
@@ -26,184 +26,210 @@ SOFTWARE.
+            // slice without fused output-layout toggle (NDHWC -> NDHWC)
+            else if((srcGenericDescPtr->layout == RpptLayout::NDHWC) && (dstGenericDescPtr->layout == RpptLayout::NDHWC))
+            {
+                srcPtrChannel = srcPtrTemp + (anchor[0] * srcGenericDescPtr->strides[1]) + (anchor[1] * srcGenericDescPtr->strides[2]) + (anchor[2] * layoutParams.bufferMultiplier);
```
L109-L121 are the same for NCDHW and NDHWC if you pass a hardcoded int3 or a tuple (0, 1, 2) for the second (1, 2, 3) for the first. Same comment for numDims = 3. 
Code for 2 and 1 seems short enough.

Same comment for HIP.

r-abishek reviewed (changes requested):
@sampath1117 Please address comments

sampath1117 commented on src/modules/rppt_tensor_geometric_augmentations.cpp:
```diff
@@ -975,43 +975,47 @@ RppStatus rppt_slice_host(RppPtr_t srcPtr,
         layoutParams = get_layout_params(srcGenericDescPtr->layout, srcGenericDescPtr->dims[4]);
-
-    if ((srcGenericDescPtr->dataType != RpptDataType::F32) && (srcGenericDescPtr->dataType != RpptDataType::U8)) return RPP_ERROR_INVALID_SRC_DATATYPE;
-    if ((dstGenericDescPtr->dataType != RpptDataType::F32) && (dstGenericDescPtr->dataType != RpptDataType::U8)) return RPP_ERROR_INVALID_DST_DATATYPE;
```
Done
Not sure why it is not reflecting in the above UI

sampath1117 commented on src/modules/hip/kernel/slice.hpp:
```diff
@@ -32,79 +87,487 @@ __global__ void slice_ncdhw_hip_tensor(T *srcPtr,
+
+    if(numDims == 4)
+    {
+        Rpp32u paramDims = numDims;
```
paramDims is not changed. Removed that variable

sampath1117 commented on utilities/test_suite/HOST/Tensor_host_audio.cpp:
```diff
@@ -109,6 +109,19 @@ int main(int argc, char **argv)
+        descriptorPtr3D->numDims = 2;
+        descriptorPtr3D->offsetInBytes = 0;
+        descriptorPtr3D->dataType = RpptDataType::F32;
+        descriptorPtr3D->dims[0] = batchSize;
```
set_generic_descriptor() function used for voxel will set numDims as 5 and will set dims and strides values accordingly
https://github.com/sampath1117/rpp/blob/sr/slice_api_changes/utilities/test_suite/rpp_test_suite_voxel.h#L168

For the audio slice case, it is just 1D input. So cannot use that function here

sampath1117 commented:
@r-abishek 
I also modified the NSR kernel output types from float to int to make it compatible with latest slice kernel

Older slice audio kernel was taking float anchor and shape as inputs which is why NSR outputs are set as float before. Now the slice kernel is accepting int anchor and shape, i changed the output types to int for NSR

r-abishek commented on src/modules/rppt_tensor_geometric_augmentations.cpp:
```diff
@@ -987,31 +990,35 @@ RppStatus rppt_slice_host(RppPtr_t srcPtr,
     if ((dstGenericDescPtr->dataType != RpptDataType::F32) && (dstGenericDescPtr->dataType != RpptDataType::U8)) return RPP_ERROR_INVALID_DST_DATATYPE;
-    if ((srcGenericDescPtr->layout != RpptLayout::NCDHW) && (srcGenericDescPtr->layout != RpptLayout::NDHWC)) return RPP_ERROR_INVALID_SRC_LAYOUT;
-    if ((dstGenericDescPtr->layout != RpptLayout::NCDHW) && (dstGenericDescPtr->layout != RpptLayout::NDHWC)) return RPP_ERROR_INVALID_DST_LAYOUT;
     if (srcGenericDescPtr->layout != dstGenericDescPtr->layout) return RPP_ERROR_INVALID_ARGUMENTS;
```
Please add an enum RPP_ERROR_LAYOUT_MISMATCH
Please change order to first check this instead of computing layout params etc.

```
if (srcGenericDescPtr->layout != dstGenericDescPtr->layout) return RPP_ERROR_LAYOUT_MISMATCH;
if ((srcGenericDescPtr->dataType != RpptDataType::F32) && (srcGenericDescPtr->dataType != RpptDataType::U8)) return RPP_ERROR_INVALID_SRC_DATATYPE;
if ((dstGenericDescPtr->dataType != RpptDataType::F32) && (dstGenericDescPtr->dataType != RpptDataType::U8)) return RPP_ERROR_INVALID_DST_DATATYPE;

RppLayoutParams layoutParams;
if (srcGenericDescPtr->layout == RpptLayout::NCDHW)
    layoutParams = get_layout_params(srcGenericDescPtr->layout, srcGenericDescPtr->dims[1]);
else if (srcGenericDescPtr->layout == RpptLayout::NDHWC)
    layoutParams = get_layout_params(srcGenericDescPtr->layout, srcGenericDescPtr->dims[4]);
```

r-abishek commented on src/modules/hip/kernel/slice.hpp:
```diff
@@ -32,77 +87,441 @@ __global__ void slice_ncdhw_hip_tensor(T *srcPtr,
+            dimsOrder[0] = 1;  // height
+            dimsOrder[1] = 2;  // width
+            int globalThreads_x = (dstGenericDescPtr->strides[2] + 7) >> 3; // W - width  (x direction) - vectorized for 8 element loads/stores per channel
+            int globalThreads_y = dstGenericDescPtr->dims[2];               // H - height (y direction)
```
There is something definitely wrong here if it works with this, since scope of the variable is within the if

r-abishek replied:
Okay the numDims = 4 is correct.
Just for numDims = 3 please remove the re-declaration of int for L200-202

r-abishek replied:
Not sure how this kernel was working with global x and global y using wrong values

r-abishek commented on src/modules/hip/kernel/slice.hpp:
```diff
@@ -32,77 +87,441 @@ __global__ void slice_ncdhw_hip_tensor(T *srcPtr,
+                                ((anchor[dimsOrder[1]] + shape[dimsOrder[1]]) > length[dimsOrder[1]]));
+
+            // launch kernel for filling the padded region with fill value specified
+            if (needPadding && dstGenericDescPtr->layout == RpptLayout::NCHW)
```
Like this condition, pick either src everywhere or dst everywhere. Since we know src layout = dst layout for this function.

sampath1117 commented on src/modules/cpu/kernel/slice.hpp:
```diff
@@ -26,184 +26,202 @@ SOFTWARE.
+            Rpp32s dimsOrder[3] = {0, 1, 2}; // for NDHWC layout
+            if((srcGenericDescPtr->layout == RpptLayout::NCDHW) && (dstGenericDescPtr->layout == RpptLayout::NCDHW))
+            {
+                dimsOrder[0] = 1;  // depth
```
Getting below error with doing initialization in single line based on condition
`initializer list cannot be used on the rig
ht hand side of operator '?'                                                                         
            Rpp32s dimsOrder[2] = (srcGenericDescPtr->layout == RpptLayout::NHWC) ? {0, 1} : {1, 2};`
  
 So changed this to regular if else case
 ```
 // order of dims
            Rpp32s dimsOrder[3];
            if (dstGenericDescPtr->layout == RpptLayout::NCDHW)
            {
                dimsOrder[0] = 1;  // depth
                dimsOrder[1] = 2;  // height
                dimsOrder[2] = 3;  // width
            }
            else
            {
                dimsOrder[0] = 0;  // depth
                dimsOrder[1] = 1;  // height
                dimsOrder[2] = 2;  // width
            }
```

sampath1117 commented:
@r-abishek 
I added slice test case in image test suite and added QA tests for the same
